### ISSUE TITLE
Audience local validation

### ISSUE THREAD

**Issue by aiscrim** [Rzpeg/IdentityServer3.AccessTokenValidation#4]
When using local validation, the ValidAudience parameter is built using the IssuerName and adding a trailing /resources to it.
This results in a SecurityTokenInvalidAudienceException being thrown, because as far as I understand, the audience is specified in the RequiredScopes, not in the Issuer field.
So I cloned the repository and added the following setting to the TokenValidationParameters and the token validation now works:

`ValidAudiences = options.RequiredScopes`

Do you think this is the correct way of handling it or I should have done something else?

**Comment by Rzpeg** [comment]
@aiscrim Audiences might not be the same thing as scopes, if I am not mistaken.
An API can have "aud" set to "api1", but have multiple scopes, like "api1.read" and "api1.write", without them being separate audiences.

**Comment by bigred8982** [comment]
The idea behind Audience is that it's the intended target of the token. In other words, the specific resource server. This can be interpreted in a few different ways and it turns out that different IDPs handle it differently.

In Identity Server 3, this is a hard-coded value, and hard coded to the issuer + "/resources". 
https://github.com/IdentityServer/IdentityServer3/blob/fd8fa2d5bed28dc38cdd869857258fea1f859935/source/Core/Services/Default/DefaultTokenService.cs#L215

The intended use of IdentityServer3 is to sit alongside the applications all on the same server so that the audience logically makes sense. "The audience for this token is https://myserver.com/resources". But, the reality is that this hard-coded value doesn't make sense if you deploy it to a different location.

What makes matters worse is this is one of the best libraries to use for validating JWT tokens when your app uses .Net Framework, regardless of which OAuth2 implementation issued the token. This is a problem if you want to validate this because the code in **this** validation library is also hard-coded to match how IdentityServer3 works.

Here's the line of code:
https://github.com/Rzpeg/IdentityServer3.AccessTokenValidation/blob/master/source/AccessTokenValidation/IdentityServerBearerTokenValidationAppBuilderExtensions.cs#L152

Audience should be fully configurable and it should be up to you to know what that value is. In IdentityServer4, Keycloak, Okta, Ping, etc, that value is configurable. Since this particular library is often used to help older application validate tokens issued from modern IDPs, I think it would be very helpful to allow this to be configurable. 

The code is short and sweet and I'll drop a PR and let you decide if you want to uptake it.

**Comment by Rzpeg** [comment]
@bigred8982 thanks for the info. 
I will gladly accept the PR. Just make sure there is no regression if no custom aud is configured.

**Comment by bigred8982** [comment]
So that's the only rub.

I originally wrote it to allow you bypass audience validation. There is a business scenario for that - at least where I've seen this library used. To support that, I would need to break backward compatibility in that if you don't configure it, it won't validate the audience at all.

But, since you are always validating the issuer, which is server name, and the audience is hard-coded to essentially match the issuer, I don't see any point in making it validate the old audience value if you haven't configured any audiences.

As an example, this is the incoming token:
```
{
  iss: "http://example.com",
  aud: "http://example.com/resources"
}
```

Since 'aud' is always hard-coded to `{iss}/resources`, it's basically redundant as the issuer value is always validated anyway.

So, I'm somewhat introducing a breaking change, but not quite. The net effect is zero impact if you haven't configured a valid audience - which pretty much everyone using this library up to now. On the plus side, you gain the benefit of being able to configure an audience value only if you need to. 

I have seen people ignoring the audience completely because they haven't quite grocked how the IDP itself is intending for you to use that value.

**Comment by Rzpeg** [comment]
@bigred8982
I mean, if no aud is configured, force the current aud value which is /resources. It's not the correct way, but it will not break existing usages.

The build is failing by the way https://ci.appveyor.com/project/Rzpeg/identityserver3-accesstokenvalidation/builds/27320888

**Comment by bigred8982** [comment]
Since the issuer is always being validating, additionally validating the hard-coded value (based off that original value) ultimately has no benefit. But, I totally understand your hesitation. And, this is your fork, after all :)

I'll change the behavior to use the default when no value is configured.

**Comment by Rzpeg** [comment]
@bigred8982 Hey.
Will you change the behavior so I could merge the request?

**Comment by Rzpeg** [comment]
@bigred8982
@NiSHoW
Will this advance?
If you need the feature, please make a pull request against a fresh master branch (IdentityModel has been updated to v4)

Edit.
I did rebase your pull request and merged it. Will publish a new version soon.

**Comment by Rzpeg** [comment]
@bigred8982 
@aiscrim 
Will be published as v 4.1.16

**Comment by aiscrim** [comment]
Cool, thanks a lot!
